bundles/sirix-core/src/main/java/org/sirix/io/file/FileStorage.java

```diff
@@ -0,0 +1,147 @@
+ * above copyright notice, this list of conditions and the following disclaimer. * Redistributions
+ * in binary form must reproduce the above copyright notice, this list of conditions and the
+ * following disclaimer in the documentation and/or other materials provided with the distribution.
+ * * Neither the name of the University of Konstanz nor the names of its contributors may be used to
+ * endorse or promote products derived from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR
+ * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND
+ * FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL <COPYRIGHT HOLDER> BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
+ * OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+ * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package org.sirix.io.file;
+
+import org.sirix.access.ResourceConfiguration;
+import org.sirix.exception.SirixIOException;
+import org.sirix.io.Reader;
+import org.sirix.io.Storage;
+import org.sirix.io.Writer;
+import org.sirix.io.bytepipe.ByteHandlePipeline;
+import org.sirix.io.bytepipe.ByteHandler;
+import org.sirix.page.PagePersister;
+import org.sirix.page.SerializationType;
+
+import java.io.IOException;
+import java.io.RandomAccessFile;
+import java.nio.file.Files;
+import java.nio.file.Path;
+
+/**
+ * Factory to provide File access as a backend.
+ *
+ * @author Sebastian Graf, University of Konstanz.
+ *
+ */
+public final class FileStorage implements Storage {
+
+  /** Data file name. */
+  private static final String FILENAME = "sirix.data";
+
+  /** Revisions file name. */
+  private static final String REVISIONS_FILENAME = "sirix.revisions";
+
+  /** Instance to storage. */
+  private final Path file;
+
+  /** Byte handler pipeline. */
+  private final ByteHandlePipeline byteHandlerPipeline;
+
+  /**
+   * Constructor.
+   *
+   * @param resourceConfig the resource configuration
+   */
+  public FileStorage(final ResourceConfiguration resourceConfig) {
+    assert resourceConfig != null : "resourceConfig must not be null!";
+    file = resourceConfig.resourcePath;
+    byteHandlerPipeline = resourceConfig.byteHandlePipeline;
+  }
+
+  @Override
+  public Reader createReader() throws SirixIOException {
+    try {
+      final Path dataFilePath = createDirectoriesAndFile();
+      final Path revisionsOffsetFilePath = getRevisionFilePath();
+
+      return new FileReader(new RandomAccessFile(dataFilePath.toFile(), "r"),
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the resource created by the following constructor call: *RandomAccessFile*. Currently, there are execution paths that do not contain closure statements, e.g., when exception is thrown by RandomAccessFile constructor. Either a) close the object returned by *RandomAccessFile()* in a try-finally block or b) close the resource by declaring the object returned by *RandomAccessFile()* in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).